### ISSUE TITLE
for redundant parens from and to fields do not match the start/end line/column in the json output.. sometimes

### ISSUE THREAD

**Issue by dten** [ndmitchell/hlint#1446]
for the following example 

```
main = do
  print $ (\((Just a), b) -> show a) (Just 1, 2)
```

hlint suggests 

```
Main.hs:2:14-21: Suggestion: Redundant bracket
Found:
  ((Just a), b)
Perhaps:
  (Just a, b)

1 hint
```

and the json output is as follows

```
[
  {
    "module": [
      "Main"
    ],
    "decl": [
      "main"
    ],
    "severity": "Suggestion",
    "hint": "Redundant bracket",
    "file": "Main.hs",
    "startLine": 2,
    "startColumn": 14,
    "endLine": 2,
    "endColumn": 22,
    "from": "((Just a), b)",
    "to": "(Just a, b)",
    "note": [],
    "refactorings": "[Replace {rtype = Pattern, pos = SrcSpan {startLine = 2, startCol = 14, endLine = 2, endCol = 22}, subts = [(\"x\",SrcSpan {startLine = 2, startCol = 15, endLine = 2, endCol = 21})], orig = \"x\"}]"
  }
]
```

the issue I'm having is that if you take the location information and replace that range in the file with the suggestion you get a bad outcome of

```
main = do
  print $ (\((Just a, b), b) -> show a) (Just 1, 2)
```

the output says the bad text is in the range 14 to 22 but the `from` text is actually from columns 13 to 28. the `to` is a valid replacement for the `from` but not the range is mismatching.

this doesn't happen with simpler redundant parens  `print $ (\(((a), b), b) -> show a) (Just 1, 2)` correctly suggests the right columns and a matching replacement

**Comment by zliu41** [comment]
Thanks for reporting the bug. I'd slightly prefer keeping the `((Just a), b)` and updating the source span, since it clearly shows why the parens are redundant.

**Comment by ndmitchell** [comment]
Agreed, updating the source span seems like the right thing to do.

**Comment by dten** [comment]
Just increasing the span would make any highlighting in editors etc less precise though. Most other lints do not give the context around them, even most other instances of redundant parens do not.

**Comment by ndmitchell** [comment]
I guess if other redundant parens don't, this should be consistent.

**Comment by cdepillabout** [comment]
Here's another example where the `from` and `to` fields do not seem to match the `startColumn` and `endColumn` values in `hlint --json`:

```haskell
main = print $ fmap (+1) $ [1,2,3]
```

Here's what `hlint` outputs for this file:

```
myfile.hs:1:16-34: Suggestion: Use <$>
Found:
  fmap (+ 1) $ [1, 2, 3]
Perhaps:
  ((+ 1) <$> [1, 2, 3])

myfile.hs:1:26: Suggestion: Redundant $
Found:
  fmap (+ 1) $ [1, 2, 3]
Perhaps:
  fmap (+ 1) [1, 2, 3]

2 hints
```

You can see that these both look correct.

However, here's what `hlint --json` outputs for this file:

```json
[
  {
    "module": [
      "Main"
    ],
    "decl": [
      "main"
    ],
    "severity": "Suggestion",
    "hint": "Use <$>",
    "file": "./myfile.hs",
    "startLine": 1,
    "startColumn": 16,
    "endLine": 1,
    "endColumn": 35,
    "from": "fmap (+ 1) $ [1, 2, 3]",
    "to": "((+ 1) <$> [1, 2, 3])",
    "note": [],
    "refactorings": "[Replace {rtype = Expr, pos = SrcSpan {startLine = 1, startCol = 16, endLine = 1, endCol = 35}, subts = [(\"f\",SrcSpan {startLine = 1, startCol = 21, endLine = 1, endCol = 25}),(\"x\",SrcSpan {startLine = 1, startCol = 28, endLine = 1, endCol = 35})], orig = \"(f <$> x)\"}]"
  },
  {
    "module": [
      "Main"
    ],
    "decl": [
      "main"
    ],
    "severity": "Suggestion",
    "hint": "Redundant $",
    "file": "./myfile.hs",
    "startLine": 1,
    "startColumn": 26,
    "endLine": 1,
    "endColumn": 27,
    "from": "fmap (+ 1) $ [1, 2, 3]",
    "to": "fmap (+ 1) [1, 2, 3]",
    "note": [],
    "refactorings": "[Replace {rtype = Expr, pos = SrcSpan {startLine = 1, startCol = 16, endLine = 1, endCol = 35}, subts = [(\"a\",SrcSpan {startLine = 1, startCol = 16, endLine = 1, endCol = 25}),(\"b\",SrcSpan {startLine = 1, startCol = 28, endLine = 1, endCol = 35})], orig = \"a b\"}]"
  }
]
```

The `startColumn` and `endColumn` values don't really seem to match up with `from` and `to`.

`endColumn` minus `startColumn` is only one character, however `from` and `to` give the whole expression.  This seems to be different from how the `startColumn` / `endColumn` / `from` / `to` values work for other lints (see the first lint for an example).
 
However, you can see that the values given in `refactorings` _do_ look correct.
